### PR TITLE
Improve mouse button parsing: tray mouse bindings

### PR DESCRIPTION
**Pull request by RedSoxFan**
~**DO NOT MERGE UNTIL AFTER #3341 and #3344**~

This is the fifth in a series of follow-up PRs for #3313.

This modifies `bar_cmd_tray_bindsym` to use `get_mouse_bindsym` for
parsing mouse buttons. This also introduces `bar_cmd_tray_bindcode`,
which will use `get_mouse_bindcode` for parsing mouse buttons. Like with
sway bindings, the two commands are encapsulated in a single file to
maximize shared code.

This also modifies tray bindings to work off of event codes rather than
x11 buttons, which allows for any mouse buttons to be used.

For `get_bar_config`, `event_code` has been added to the `tray_bindings`
section and will include the event code for the button. If the event code
can be mapped to a x11 button, `input_code` will still be the x11 button
number. Otherwise, `input_code` will be `0`.

### CODE REVIEW

**Comment by RedSoxFan** [comment]
Rebased and ready for review

**Comment by RedSoxFan** [comment]
Actually hold off on this until after #3403 gets merged and I'll rebase and alter the calls to `cmd_results_new`.

**Comment by RedSoxFan** [comment]
Rebased and removed the `input` argument for the `cmd_results_new` calls.

This is ready for review.

File: include/sway/config.h [hunk 1]
```diff
@@ -272,6 +272,13 @@ struct bar_binding {
 	char *command;
 };
 
+#if HAVE_TRAY
+struct tray_binding {
+	uint32_t button;
+	const char* command;
```
Comment (ianyfan) [inline include/sway/config.h]: Style: `char *command`

File: sway/commands/bar/tray_bind.c [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+static struct cmd_results *tray_bind(int argc, char **argv, bool code) {
+#if HAVE_TRAY
+	const char *command = code ? "bar tray_bindcode" : "bar tray_bindsym";
+	struct cmd_results *error = NULL;
+	if ((error = checkarg(argc, command, EXPECTED_EQUAL_TO, 2))) {
+		return error;
+	}
+	if (!config->current_bar) {
+		return cmd_results_new(CMD_FAILURE, command, "No bar defined.");
```
Comment (ianyfan) [inline sway/commands/bar/tray_bind.c]: I don't think this is correct?

File: include/sway/config.h (hunk 1 above)
Comment (RedSoxFan) [reply include/sway/config.h]: Fixed

File: sway/commands/bar/tray_bind.c (hunk 2 above)
Comment (RedSoxFan) [reply sway/commands/bar/tray_bind.c]: Oops, missed that one. Fixed.

File: include/sway/config.h [hunk 3]
```diff
@@ -260,7 +260,7 @@ struct bar_config {
 
 #if HAVE_TRAY
 	char *icon_theme;
-	const char *tray_bindings[10]; // mouse buttons 0-9
+	list_t *tray_bindings; // struct tray_binding
```
Comment (ianyfan) [inline include/sway/config.h]: `struct tray_binding *`

Comment (ianyfan) [inline include/sway/config.h]: `struct tray_binding *`

File: include/swaybar/config.h [hunk 4]
```diff
@@ -70,13 +70,20 @@ struct swaybar_config {
 
 #if HAVE_TRAY
 	char *icon_theme;
-	char *tray_bindings[10]; // mouse buttons 0-9
+	list_t *tray_bindings; // struct tray_binding
```
Comment (ianyfan) [inline include/swaybar/config.h]: `struct tray_binding *`

File: include/sway/config.h (hunk 3 above)
Comment (RedSoxFan) [reply include/sway/config.h]: Changed

File: include/swaybar/config.h (hunk 4 above)
Comment (RedSoxFan) [reply include/swaybar/config.h]: Changed

File: include/sway/config.h [hunk 5]
```diff
@@ -260,7 +260,7 @@ struct bar_config {
 
 #if HAVE_TRAY
 	char *icon_theme;
-	const char *tray_bindings[10]; // mouse buttons 0-9
+	list_t *tray_bindings; // struct tray_binding *
```
Comment (ddevault) [inline include/sway/config.h]: How about wl_list instead?

Comment (RedSoxFan) [reply include/sway/config.h]: Uh sure, I can change it to use wl_list. I just used list_t since that's what all the other bindings are using

Comment (ddevault) [reply include/sway/config.h]: Yeah, I should have caught this in the other patches. We should be moving away from list_t.

**Comment by emersion** [comment]
Thanks!

**Pull request merged**